rfcs/45-stageless.md

```diff
@@ -0,0 +1,761 @@
+    population_limit: Res<PopulationLimit>,
+) -> bool {
+    population_query.iter().count() > population_limit
+}
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        // You can add functions with read-only system parameters as conditions.
+        .add_set(GameSet::Construction.run_if(construction_timer_not_finished))
+        .add_many(systems![tick_construction_timer, update_construction_progress].in_set(GameSet::Construction))
+        .add_system(system_meltdown_mitigation.run_if(too_many_enemies))
+        // We can use closures for simple one-off conditions, 
+        // which automatically fetch the appropriate data from the `World`.
+        .add_system(spawn_more_enemies.run_if(|difficulty: Res<Difficulty>| difficulty >= 9000))
+        // The `resource_exists` and `resource_equals` helper functions for generating
+        // simple conditions (well, closures that convert into conditions).
+        .add_system(gravity.run_if(resource_exists(Gravity)))
+        // Conditions can be attached to system sets, so you can skip the whole set.
+        .add_set(GameSet::Physics.run_if(resource_equals(Paused(false))))
+        .run();
+}
+```
+
+### Running scheduled systems
+
+The systems, sets, and scheduling data added to an `App` is stored on the `World` in the `SystemRegistry` resource. This resource is also responsible for generating and maintaining the dependency graphs for the stored system sets.
+
+In order to run a system or system set on the `World`, the system(s) must be extracted from the `SystemRegistry`. (This lets the `SystemRegistry` remain accessible by the running systems).
+
+To do that, the `SystemRegistry` has methods to export a system or schedule given a label.
+A **schedule** is an executable version of a system set, containing all the systems and conditions, as well as some metadata for an **executor** to run those in the correct order (efficiently).
+After the schedule has been run, it can be returned to the `SystemRegistry` using another method.
+This is the same process the default `App` runner uses to execute the startup sequence and main update loop.
+
+### Command processing
+
+Commands are arbitrary world modifications that are most commonly used to spawn or despawn entities and add or remove components.
+Since those types of changes have unpredictable side effects on stored component data, commands are deferred until the next scheduled `apply_buffers` exclusive system runs.
+
+```rust
+use bevy::prelude::*;
+
+struct ProjectilePlugin;
+
+impl Plugin for ProjectilePlugin {
+    fn build(app: &mut App) {
+        app
+        .add_system(spawn_projectiles.before(CoreSystem::FlushPostUpdate))
+        .add_many(
+            chain![
+                check_if_projectiles_hit,
+                despawn_projectiles_that_hit,
+                // Wherever you want to sync and apply commands, insert an instance of `apply_buffers`.
+                apply_buffers.named("FlushProjectiles")

```

Comment:
Wouldn't it be better to have a macro here, so users explicit know that `apply_buffers` is a built-in system? Something like `apply_buffers!("FushProjectiles")`.

My concern is letting it explicit to users that this system is an exclusive system and may cause "hard sync points" AFAIK.

Reply:
How would turning it into a macro help communicate that?

Reply:
Maybe there can be a better way, like `core_apply_buffers`?

Anyway having a macro is an option to call user attention to that point, because it's something different.

But may not be the best way